### ISSUE TITLE
Improve 'json-prune' — match key with specific value, remove array or object if it contains specific item

### ISSUE THREAD

**Issue by slavaleleka** [AdguardTeam/Scriptlets#183]
there are few issues:

1  current json-prune does not match values if we need specify pruning, only keys — `obligatoryProps` — are checked 

2  if specific array item has to be pruned and we can target such items by their extra fields
```
#%#//scriptlet('json-prune', 'recs_group.[].tiles.[].[].*', 'recs_group.[].tiles.[].[].ad_origin')
```
but whole `recs_group.[].tiles.[].[]` array items will be pruned, not just ones with `ad_origin`.
check on [fontanka.ru](https://www.fontanka.ru/2018/11/30/011/)
https://uploads.adguard.com/slbdzdrv7cfxo.png

**Comment by ameshkov** [comment]
@slavaleleka what do you think about implementing a scriptlet that supports `jq`-like syntax for that? At least a subset of it.
https://stedolan.github.io/jq/tutorial/

Alternatively, there is this: https://jmespath.org/

I do not suggest supporting it fully, but we could take that syntax as a base and build our limited subset of it. This way it'd be easier for the scriptlet users to understand the capabilities.

PS: we'll probably need to reassign this task to a later milestone.

**Comment by ngorskikh** [comment]
@ameshkov How about this for a rule spec:

### `$jsonprune` modifier

`$jsonprune` rules modify the JSON response of a matching request by removing JSON items that match a modified (see below) 
[JSONPath](https://goessner.net/articles/JsonPath/) expression. They do not modify responses which are not valid JSON.

### Syntax
* `||example.org^$jsonprune=expression` – remove items that match the modified JSONPath expesssion `expression` from the response.

Due to the way rule parsing works, the characters `$` and `,` must be escaped with `\` inside `expression`.

The modified JSONPath syntax has the following differences from the original:
1. Script expressions are not supported.
2. The supported filter expressions are:
2.1. `?(has <key>)` -- true if the current object has the specified key.
2.2. `?(key-eq <key> <value>)` -- true if the current object has the specified key,
                                  and its value is equal to the specified value.
3. Whitespace outside of double- or single-quoted strings has no meaning.
4. Both double- and single-quoted strings can be used.
5. Expressions ending with `..` are not supported.
6. Multiple array slices can be specified in square brackets.

### Exceptions

Basic URL exceptions shall not disable `$jsonprune` rules. They can be disabled as described below:
* `@@||example.org^$jsonprune` – disable all `$jsonprune` rules for responses from URLs matching `||example.org^`.
* `@@||example.org^$jsonprune=text` – disable all `$jsonprune` rules with the value of the `jsonprune` modifier equal to `text` for responses from URLs matching `||example.org^`.
* `$jsonprune` rules can also be disabled by `$document`, `$content` and `$urlblock` exception rules.

### Restrictions
* `$jsonprune` rules are only allowed in trusted filters.
* `$jsonprune` rules are not compatible with any other modifiers except `$domain`, `$third-party`, `$app`, `$important`, `$match-case`, and `$xmlhttprequest`.
* `$jsonprune` rules do not apply if the size of the original response is more than 3 MB.

### Notes
* When multiple `$jsonprune` rules match the same request, they are sorted in lexicographical order,
the first rule is applied to the original response, and each of the remaining rules is applied
to the result of applying the previous one.

### Examples
* `||example.org^$jsonprune=\$..[one\, "two three"]` — remove all occurences of the keys "one" and "two three" anywhere in the JSON document.
* `||example.org^$jsonprune=\$.a[?(has ad_origin)]` – remove all children of `a` that have an `ad_origin` key.
* `||example.org^$jsonprune=\$.*.*[?(key-eq 'Some key' 'Some value')]` – remove all items that are at nesting level 3 and have a property "Some key" equal to "Some value".

**Comment by ameshkov** [comment]
@ngorskikh the spec should've been posted here I think: https://github.com/AdguardTeam/CoreLibs/issues/1447

We'll also need a filter expression for checking whether the value contains the specified string: `?(key-contains <key> <value>)`

**Comment by piquark6046** [comment]
Currently, `$jsonprune` modifier can modify only response of a request.
However, during maintaining List-KR filter, I often need its scriptlet version.

Example:
<details>
<summary>Gmarket.co.kr</summary>

URL: https://gmarket.co.kr/
Description: Extended CSS can hide the elements having `AD` label. However, their network requests cannot be blocked with the previous modifiers and URL rules (including a RegExp)
Related Issue: https://github.com/List-KR/List-KR/issues/563
Targeted array items: `$props.pageProps.serverData.topWideBannerSection.banners[?(@.isAd == true)]`

</details>

**Comment by ameshkov** [comment]
Could you please explain what you're trying to do with that request?

Do you need to block a request that contains such a JSON path?

**Comment by piquark6046** [comment]
https://github.com/AdguardTeam/Scriptlets/issues/183#issuecomment-1336472219
Oh, I forgot to add what i add.
What i want is a feature removing a property with a JSON path from result of calling `JSON.parse`.

**Comment by ameshkov** [comment]
So you'd like a scriptlet alternative to `$jsonprune`?

**Comment by ameshkov** [comment]
Makes sense, `$jsonprune` syntax is more powerful and the end goal was indeed to provide a scriptlet version of it.

Moved it to Scriptlets v1.9 for now. @slavaleleka @stanislav-atr what do you think about it?

**Comment by slavaleleka** [comment]
we shall return to the issue later

**Comment by Alex-302** [comment]
`Later` is now, @slavaleleka  :)

**Comment by slavaleleka** [comment]
the same is needed is scriptlets as well:
https://github.com/AdguardTeam/CoreLibs/issues/1717

**Comment by slavaleleka** [comment]
[second issue in first issue post](https://github.com/AdguardTeam/Scriptlets/issues/183#:~:text=2%20if%20specific%20array%20item%20has%20to%20be%20pruned%20and%20we%20can%20target%20such%20items%20by%20their%20extra%20fields) cannot be fixed in current syntax implementation — `obligatoryProps` is checked **for the whole json** and there is no way to check exact inner array item child (which is object) and conditionally prune it while other array items remains unchanged

**Comment by ameshkov** [comment]
@slavaleleka not sure what you mean, aren't filter expressions supposed to solve this?

**Comment by slavaleleka** [comment]
I mean _current syntax implementation_, not jsonpath filter expression

**Comment by maximtop** [comment]
I couldn't figure out how we can handle JSONP requests, apart from one method. For example, let's say there's a JSONP callback like this:

```html
...
<script>
    function handleResponse(data) {
        console.log(JSON.stringify(data));
    }
</script>
<script src="http://server.com/jsonp?callback=handleResponse"></script>
...
```
We can find the matching callback function from the page and proxy the call to it.

**Comment by ameshkov** [comment]
Probably jsonp shouldn't be in scope of this issue since there's no clear solution.

**Comment by piquark6046** [comment]
Please take a look for json-prune scriptlet version: https://github.com/gorhill/uBlock/commit/3152896d428c54c76cfd66c3da110bd4d6506cbc and https://github.com/gorhill/uBlock/commit/14d60ac5d3a1c5c2b9b8c062ea31e4ddcd13a98f.

**Comment by Alex-302** [comment]
Another case
https://github.com/AdguardTeam/AdguardFilters/issues/195194
@maximtop
